tensorflow/core/lib/random/random_distributions.h

```diff
@@ -23,22 +23,60 @@ limitations under the License.
 #include <algorithm>
 #include <type_traits>
 
-#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor.h"
 #include "tensorflow/core/lib/bfloat16/bfloat16.h"
 #include "tensorflow/core/lib/random/philox_random.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
 
 namespace tensorflow {
 namespace random {
 
 // Helper function to convert a 16-bit integer to a half between [0..1).
 PHILOX_DEVICE_INLINE Eigen::half Uint16ToHalf(uint16 x);
+// Helper function to convert a 16-bit integer to a bfloat16 between [1..2).
+PHILOX_DEVICE_INLINE bfloat16 InternalUint16ToBfloat16(uint16 x);
 // Helper function to convert a 16-bit integer to a bfloat16 between [0..1).
-PHILOX_DEVICE_INLINE bfloat16 Uint16ToGfloat16(uint16 x);
+PHILOX_DEVICE_INLINE bfloat16 Uint16ToBfloat16(uint16 x);
+// Helper function to convert a 32-bit integer to a float between [1..2).
+PHILOX_DEVICE_INLINE float InternalUint32ToFloat(uint32 x);
 // Helper function to convert a 32-bit integer to a float between [0..1).
 PHILOX_DEVICE_INLINE float Uint32ToFloat(uint32 x);
 // Helper function to convert two 32-bit integers to a double between [0..1).
 PHILOX_DEVICE_INLINE double Uint64ToDouble(uint32 x0, uint32 x1);
 
+// Helper function to format distribution result in vectorization path,
+// it creates Eigen::Tensor and reuses packet feature with SIMD.
+template <typename Distribution, typename Generator>
+typename Distribution::ResultType VectorizedFormat(
+    Generator* gen, typename Distribution::FormatFunc functor) {
+  typename Generator::ResultType sample;
+  typename Distribution::ResultType result;
+  const int kResultElementCount = Distribution::kResultElementCount;
+  const int inner_count = Generator::kResultElementCount;
+  const int outer_count = kResultElementCount / inner_count;
+  int offset = 0;
+
+  for (int k = 0; k < outer_count; k++) {
+    sample = (*gen)();
+    for (int i = 0; i < inner_count; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+  // Tail processing if any.
+  if (offset < kResultElementCount) {
+    sample = (*gen)();
+    for (int i = 0; offset < kResultElementCount; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+
+  auto tensor_result =
+      typename TTypes<typename Distribution::ResultElementType>::Tensor(
+          &result[0], kResultElementCount);
+  tensor_result = tensor_result - typename Distribution::ResultElementType(1.0);
```

Comment:
The original `minus` is not in `gen`, it's the format function in the loop of `Distribution`: https://github.com/tensorflow/tensorflow/blob/2b2df19f4b2a249821611e98786262585b57bebe/tensorflow/core/lib/random/random_distributions.h#L126.

Key optimization of this PR is to move the `minus` out of the loop so that it can be vectorized.

Reply:
I see it now. I think I misread between `Uint16toBfloat16/InternalUint16toBfloat16` and `Uint32toFloat/InternalUint32toFloat` at the time.